File: reviews/python/classification.py
#!/usr/bin/python3

import numpy as np
import re, string, pickle
from sklearn.datasets import load_files
from datetime import datetime

# Remove html from html and turn lowercase
def custom_standardization(input_data):
    lowercase = input_data.lower()
    stripped_html = re.sub("<br />", " ", str(lowercase))
    return re.sub(
         f"[{re.escape(string.punctuation)}]", "", stripped_html
    )

# Load text classifier
name = 'text_classifier'
print(datetime.now(), f'Loading model {name}')
with open(name, 'rb') as training_model:
    model = pickle.load(training_model)  

# Load text vectorizer    
name = 'text_vectorizer'
print(datetime.now(), f'Loading model {name}')
with open(name, 'rb') as vect:
    vectorizer = pickle.load(vect)      

while True:
    x = input(str(datetime.now()) + ' Write you review below!\n> ')
    # Vectorize input
    x = vectorizer.transform([custom_standardization(x)]).toarray()    
    # Classify review
    y = model.predict(x)
    
    if y == [1]:
        print('Review is positive!')
    else:    
        print('Review is negative!')

File: reviews/R/classification.r
#!/usr/bin/R -s -f

# to create a model run model.r before this script

library(tensorflow)
library(keras)
library(tfdatasets)

# creating a regex with all punctuation characters for replacing.
re <- reticulate::import("re")

punctuation <- c("!", "\\", "\"", "#", "$", "%", "&", "'", "(", ")", "*",
"+", ",", "-", ".", "/", ":", ";", "<", "=", ">", "?", "@", "[",
"\\", "\\", "]", "^", "_", "`", "{", "|", "}", "~")

punctuation_group <- punctuation %>%
  sapply(re$escape) %>%
  paste0(collapse = "") %>%
  sprintf("[%s]", .)

custom_standardization <- function(input_data) {
  lowercase <- tf$strings$lower(input_data)
  stripped_html <- tf$strings$regex_replace(lowercase, '<br />', ' ')
  tf$strings$regex_replace(
    stripped_html,
    punctuation_group,
    ""
  )
}

# loading model
model <- load_model_tf('model')


while (1) {
    cat("Write your review below!\n> ")
    x <- readLines(con = "stdin", n = 1)
    # standerdize review
    x <- custom_standardization(x)
    # predict
    y <- predict(model, as.array(x))
    if (y[1] <= 0.5){
        cat('Review is negative!\n')
    } else {
        cat('Review is positive!\n')
    }
}